docs/versioning.md

```diff
@@ -0,0 +1,463 @@
+# Versioning
+
+This section describes S3 Server's support for the specific AWS S3 Bucket
+Versioning feature.
+
+## AWS S3 Bucket Versioning
+
+An object in a versioning enabled bucket can have multiple versions, whereas an
+object can have only a single version in a standard bucket without versioning.
+
+With the standard S3 API, any update (PUT, DELETE) to an object in a versioning
+enabled bucket would create a new version of that object. Read operations such
+as GET and LIST would see the latest version of the objects. The table below
+shows such example, a `PUT foo` generates a new version `v2` and keeps `v1`;
+GET and LIST return the latest version of `foo`, which is `v2` in this case.
+
+| key | version | value |
+|-----|:-------:|:-----:|
+| foo |   v2    |   B   |
+| foo |   v1    |   A   |
+
+To work with all existing versions of an object, S3 provides an extended
+API which uses the same operations of the standard API, but with versioning
+information specified. For example, in order to get the version `v1` of `foo`,
+users have to specify that version in the query, such as
+`GET /foo?versionId=v1`.
+
+A bucket can be non-versioning or versioning. Any objects placed in a bucket
+before versioning is enabled are considered to be the "null" version. When
+versioning is enabled, the "null" version will be kept and continue to be
+identified as the "null" version. However, if versioning is suspended, the
+"null" version may be overwritten with a new "null" version. All versions with
```

Comment:
It depends on the perspective you look at it. Without knowing the details of implementation and from the perspective of someone using the S3 api, we want them to perceive what happens simply as overwriting the null version. (You can access the new version with the same version ID, it just also becomes the latest version if the null version was not the latest version before)

Although behind the scenes what happens is sometimes the null version is overwritten, other times the old version is deleted and a new version is created, and we're trying to describe what happens in this document, since we're at the overview phase here and haven't gotten to details of implementation, does it make sense to stick to the current wording?